quic/s2n-quic-core/src/task/cooldown.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use pin_project_lite::pin_project;
+
+#[derive(Clone, Debug, Default)]
+pub struct Cooldown {
+    credits: u16,
+    limit: u16,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Outcome {
+    /// The task should loop
+    Loop,
+    /// The task should return Pending and wait for an actual wake notification
+    Sleep,
+}
+
+impl Outcome {
+    #[inline]
+    pub fn is_loop(&self) -> bool {
+        matches!(self, Self::Loop)
+    }
+
+    #[inline]
+    pub fn is_sleep(&self) -> bool {
+        matches!(self, Self::Sleep)
+    }
+}
+
+impl Cooldown {
+    #[inline]
+    pub fn new(limit: u16) -> Self {
+        Self {
+            limit,
+            credits: limit,
+        }
+    }
+
+    #[inline]
+    pub fn state(&self) -> Outcome {
+        if self.credits > 0 {
+            Outcome::Loop
+        } else {
+            Outcome::Sleep
+        }
+    }
+
+    /// Notifies the cooldown that the poll operation was ready
+    ///
+    /// This resets the cooldown period until another `Pending` result.
+    #[inline]
+    pub fn on_ready(&mut self) {
+        // reset the pending count
+        self.credits = self.limit;
+    }
+
+    /// Notifies the cooldown that the poll operation was pending
+    ///
+    /// This consumes a cooldown credit until they are exhausted at which point the task should
+    /// sleep.
+    #[inline]
+    pub fn on_pending(&mut self) -> Outcome {
+        if self.credits > 0 {
+            self.credits -= 1;
+            return Outcome::Loop;
+        }
+
+        Outcome::Sleep
+    }
+
+    #[inline]
+    pub fn on_pending_task(&mut self, cx: &mut core::task::Context) -> Outcome {
+        let outcome = self.on_pending();
+
+        if outcome.is_loop() {
+            cx.waker().wake_by_ref();
+        }
+
+        outcome
+    }
+
+    #[inline]
+    pub async fn wrap<F>(&mut self, fut: F) -> F::Output
+    where
+        F: Future + Unpin,
+    {
+        Wrapped {
+            fut,
+            cooldown: self,
+        }
+        .await
+    }
+}
+
+pin_project!(
+    struct Wrapped<'a, F>
+    where
+        F: core::future::Future,
+    {
+        #[pin]
+        fut: F,
+        cooldown: &'a mut Cooldown,
+    }
+);
+
+impl<'a, F> Future for Wrapped<'a, F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let this = self.project();
+        match this.fut.poll(cx) {
+            Poll::Ready(v) => {
+                this.cooldown.on_ready();
+                Poll::Ready(v)
+            }
+            Poll::Pending => {
+                this.cooldown.on_pending_task(cx);
+                Poll::Pending
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn cooldown_test() {
```

Comment:
can you add a test that uses `Cooldown::new(0)`, since that's how customers not using the unstable feature would end up using it?